# Clase 1

# DataFrame
Un DataFrame en Python es una estructura de datos bidimensional proporcionada por la biblioteca Pandas. Se asemeja a una hoja de cálculo de Excel o a una tabla en una base de datos, donde los datos se organizan en filas y columnas.

## Características principales:
- Filas y columnas: Cada columna puede contener un tipo de dato diferente (números, cadenas, etc.), y cada fila representa una instancia de datos.
- Etiquetas: Tanto las filas como las columnas pueden tener etiquetas, lo que facilita la manipulación y el acceso a los datos.
- Métodos y atributos: Pandas ofrece una variedad de métodos para importar, manipular y analizar datos, como read_csv() para importar datos desde un archivo CSV, info() para obtener información sobre el DataFrame, y describe() para obtener estadísticas descriptivas.
- Para crear un dataframe en python se debe hacer uso de la libreria pandas. 

## ¿Qué es una librería en Python?
Una librería en Python es una colección de módulos y paquetes que contienen funciones, clases y métodos predefinidos. Estas librerías extienden la funcionalidad básica del lenguaje, permitiendo a los desarrolladores reutilizar código y acceder a funcionalidades adicionales sin tener que escribir todo desde cero. Por ejemplo, hay librerías para manipulación de datos, visualización, aprendizaje automático, y mucho más.

## ¿Qué es Pandas?
Pandas es una de las librerías más populares en Python para la manipulación y análisis de datos. Fue creada por Wes McKinney en 2008 y proporciona estructuras de datos poderosas y flexibles, como Series y DataFrames, que facilitan el trabajo con datos tabulares (similares a las hojas de cálculo de Excel).

Para usar la libreria pnadas en python debemos incluir el siguiente código
```python
import pandas as pd
```




## Ejemplo básico
```python
# Cargando la libreria pandas
import pandas as pd

# Crear un DataFrame a partir de un diccionario
data = {
    'Nombre': ['Ana', 'Luis', 'María'],
    'Edad': [23, 45, 34],
    'Ciudad': ['Madrid', 'Barcelona', 'Valencia']
}

df = pd.DataFrame(data)

print(df)

```

In [2]:
# Cargando la libreria pandas
import pandas as pd

# Crear un DataFrame a partir de un diccionario
data = {
    'Nombre': ['Ana', 'Luis', 'María'],
    'Edad': [23, 45, 34],
    'Ciudad': ['Madrid', 'Barcelona', 'Valencia']
}

df = pd.DataFrame(data)

print(df)


  Nombre  Edad     Ciudad
0    Ana    23     Madrid
1   Luis    45  Barcelona
2  María    34   Valencia


## Crea un dataframe con los datos de las personas del curso, las columnas son: nombre, apellido, edad, mascota. La columna mascota es booleano y depende de si la persona tiene mascota o no.

## Datraframe a partir de un diccionario
Tu diccionario debe tener el siguiente formato:

Las claves del diccionario serán los nombres de tus columnas.

Los valores asociados a cada clave deben ser listas (o arreglos de NumPy) que contengan los datos para esa columna específica. Es crucial que todas estas listas tengan la misma longitud, ya que cada elemento en la misma posición de cada lista formará una fila del DataFrame.

## Ejemplo

```python
import pandas as pd
datos_alumnos = {
    'Nombre': ['Ana', 'Juan', 'María', 'Pedro'],
    'Edad': [20, 22, 21, 23],
    'Carrera': ['Ingeniería', 'Derecho', 'Medicina', 'Arquitectura'],
    'Promedio': [8.5, 7.9, 9.1, 8.2]
}

```

In [3]:
import pandas as pd
datos_alumnos = {
    'Nombre': ['Ana', 'Juan', 'María', 'Pedro'],
    'Edad': [20, 22, 21, 23],
    'Carrera': ['Ingeniería', 'Derecho', 'Medicina', 'Arquitectura'],
    'Promedio': [8.5, 7.9, 9.1, 8.2]
}
# Creamos el DataFrame a partir del diccionario
df_alumnos = pd.DataFrame(datos_alumnos)

# Mostramos el DataFrame para ver su contenido
print(df_alumnos)

  Nombre  Edad       Carrera  Promedio
0    Ana    20    Ingeniería       8.5
1   Juan    22       Derecho       7.9
2  María    21      Medicina       9.1
3  Pedro    23  Arquitectura       8.2


Observa cómo pandas ha interpretado automáticamente:

Las claves del diccionario como los nombres de las columnas.

Las listas como los datos de cada columna.

Ha añadido un índice numérico por defecto (0, 1, 2, 3) para las filas.



## Otro ejemplo

```python
import pandas as pd

# Diccionario con información de productos
datos_productos = {
    'Producto': ['Laptop', 'Mouse', 'Teclado', 'Monitor'],
    'Precio': [1200, 25, 75, 300],
    'Stock': [50, 200, 150, 30],
    'Categoria': ['Electrónica', 'Accesorios', 'Accesorios', 'Electrónica']
}

# Crear el DataFrame
df_productos = pd.DataFrame(datos_productos)

# Imprimir el DataFrame
print(df_productos)
```

In [4]:
import pandas as pd

# Diccionario con información de productos
datos_productos = {
    'Producto': ['Laptop', 'Mouse', 'Teclado', 'Monitor'],
    'Precio': [1200, 25, 75, 300],
    'Stock': [50, 200, 150, 30],
    'Categoria': ['Electrónica', 'Accesorios', 'Accesorios', 'Electrónica']
}

# Crear el DataFrame
df_productos = pd.DataFrame(datos_productos)

# Imprimir el DataFrame
print(df_productos)

  Producto  Precio  Stock    Categoria
0   Laptop    1200     50  Electrónica
1    Mouse      25    200   Accesorios
2  Teclado      75    150   Accesorios
3  Monitor     300     30  Electrónica


# Data frame a partir de una lista

In [5]:
import pandas as pd

datos = [
    ['Ana', 23, 'Madrid'],
    ['Luis', 34, 'Barcelona'],
    ['Carlos', 45, 'Valencia']
]

columnas = ['Nombre', 'Edad', 'Ciudad']

df = pd.DataFrame(datos, columns=columnas)
print(df)


   Nombre  Edad     Ciudad
0     Ana    23     Madrid
1    Luis    34  Barcelona
2  Carlos    45   Valencia


# Data frame a partir de una lista de diccionarios

In [6]:
datos = [
    {'Nombre': 'Ana', 'Edad': 23, 'Ciudad': 'Madrid'},
    {'Nombre': 'Luis', 'Edad': 34, 'Ciudad': 'Barcelona'},
    {'Nombre': 'Carlos', 'Edad': 45, 'Ciudad': 'Valencia'}
]

df = pd.DataFrame(datos)
print(df)


   Nombre  Edad     Ciudad
0     Ana    23     Madrid
1    Luis    34  Barcelona
2  Carlos    45   Valencia


# Dataframe a patir de una lista simple


In [7]:
datos = ['Ana', 'Luis', 'Carlos']

df = pd.DataFrame(datos, columns=['Nombre'])
print(df)


   Nombre
0     Ana
1    Luis
2  Carlos


# Crear dataframe a partir de un archivo
### DataFrame a partir de un archivo CSV

```python
import pandas as pd

# Leer un archivo CSV y crear un DataFrame
df = pd.read_csv('archivo.csv')
print(df)
```

### Ejemplo carga el archivo datos.csv

In [8]:
import pandas as pd

# Cargar el archivo CSV
df1 = pd.read_csv('datos.csv')

# Mostrar el DataFrame
print(df1)

  Nombre  Edad   Curso  Nota
0    Ana    21  Física   6.5
1   Luis    22  Física   5.8
2  Carla    20  Física   7.2
3  Tomás    23  Física   4.9


### DataFrame a partir de un archivo Excel

```python
import pandas as pd

# Leer un archivo Excel y crear un DataFrame
df = pd.read_excel('datos.xlsx')
print(df)
```


In [9]:
import pandas as pd

# Leer un archivo Excel y crear un DataFrame
df = pd.read_excel('datos.xlsx')
print(df)


  Nombre  Edad   Curso  Nota
0    Ana    21  Física   6.5
1   Luis    22  Física   5.8
2  Carla    20  Física   7.2
3  Tomás    23  Física   4.9


# 📊 Métodos comunes de DataFrames en Pandas (con ejemplos)

## 1. `.head(n)`
**Muestra las primeras `n` filas del DataFrame.**
```python
df.head(3)
```

## 2. `.tail(n)`
**Muestra las últimas `n` filas del DataFrame.**
```python
df.tail(2)
```

## 3. `.info()`
**Muestra un resumen del DataFrame: columnas, tipos de datos, valores no nulos.**
```python
df.info()
```

## 4. `.describe()`
**Estadísticas básicas (media, desviación, min, max...) de columnas numéricas.**
```python
df.describe()
```

## 5. `.shape`
**Devuelve una tupla con el número de filas y columnas.**
```python
df.shape  # (filas, columnas)
```

## 6. `.columns`
**Devuelve el listado de nombres de columnas.**
```python
df.columns
```

## 7. `.dtypes`
**Devuelve el tipo de datos de cada columna.**
```python
df.dtypes
```

## 8. `.isnull()`
**Indica con True/False si un valor es nulo.**
```python
df.isnull()
```

## 9. `.dropna()`
**Elimina filas con valores nulos.**
```python
df.dropna()
```

## 10. `.fillna(valor)`
**Rellena los valores nulos con `valor`.**
```python
df.fillna(0)
```

## 11. `.sort_values(by='col')`
**Ordena el DataFrame según una columna.**
```python
df.sort_values(by='Edad')
```

## 12. `.groupby('col')`
**Agrupa datos por una columna (útil para operaciones estadísticas).**
```python
df.groupby('Ciudad').mean()
```

## 13. `.value_counts()`
**Cuenta la frecuencia de cada valor en una columna.**
```python
df['Ciudad'].value_counts()
```

## 14. `.loc[filas, columnas]`
**Accede a filas/columnas por etiqueta.**
```python
df.loc[0, 'Nombre']  # primer nombre
```

## 15. `.iloc[fila, columna]`
**Accede por posición (índice entero).**
```python
df.iloc[0, 1]  # valor en fila 0, columna 1
```

## 16. `.rename(columns={'ViejoNombre': 'NuevoNombre'})`
**Renombra columnas.**
```python
df.rename(columns={'Edad': 'Años'})
```

## 17. `.drop('col', axis=1)`
**Elimina una columna.**
```python
df.drop('Ciudad', axis=1)
```

## 18. `.apply(func)`
**Aplica una función a cada valor de una columna.**
```python
df['Edad'].apply(lambda x: x + 1)
```

## 19. `.astype(tipo)`
**Cambia el tipo de datos de una columna.**
```python
df['Edad'] = df['Edad'].astype(float)
```

## 20. `.duplicated()`
**Detecta filas duplicadas.**
```python
df.duplicated()
```

In [10]:
import pandas as pd
df = pd.read_csv('articulos.csv')
df

ID_Articulo                         Nombre_Articulo       Categoria  \
0              1     Altavoz Inteligente Amazon Echo Dot     Electrónica   
1              2          Anillo de Compromiso Blue Nile            Moda   
2              3                          Tostadora Smeg  Hogar y Cocina   
3              4           Termostato Inteligente Ecobee     Electrónica   
4              5                     Microscopio AmScope        Juguetes   
..           ...                                     ...             ...   
195          196             Auriculares Sony WH-1000XM5     Electrónica   
196          197                   Tablet Apple iPad Air     Electrónica   
197          198  Reloj Inteligente Apple Watch Series 9     Electrónica   
198          199           Smartphone Samsung Galaxy S24     Electrónica   
199          200                  Laptop HP Spectre x360     Electrónica   

     Precio_Unitario  Stock Disponibilidad Fecha_Ultima_Venta  \
0             552.12    391       En Stock         2025-01-28   
1             440.09    242       En Stock         2024-08-01   
2             224.23    261       En Stock         2024-11-20   
3            1359.81    215       En Stock         2025-05-19   
4             114.73    472       En Stock         2024-12-07   
..               ...    ...            ...                ...   
195          1346.06    478       En Stock         2024-11-06   
196          1479.92    109       En Stock         2024-11-06   
197          1359.81    215       En Stock         2025-05-19   
198          1404.99    149       En Stock         2025-04-20   
199           128.53    197       En Stock         2024-11-23   

     Unidades_Vendidas_Ultimo_Mes  Valor_Total_Ventas_Ultimo_Mes  
0                             114                       14275.98  
1                              92                        4427.35  
2                             131                       10012.87  
3                             178                       14169.57  
4                             175                       18854.76  
..                            ...                            ...  
195                           177                       10996.65  
196                           177                       10996.65  
197                           178                       14169.57  
198                           131                        2343.68  
199                           170                       16632.18  

[200 rows x 9 columns]

# Clase 2: Series de Panda (pandas Series)
Una serie en Pandas es una estructura unidimensional que puede contener cualquier tipo de dato (enteros, cadenas, flotantes, etc.). Es similar a una lista o un arreglo de  pero con etiquetas (índices) para cada elemento. Podemos pensar en una serie como una columna de un DataFrame, pero con la ventaja de que puede ser manipulada de forma independiente.

## Ejemplo básico de una serie pandas a partir de una lista
```python
import pandas as pd

# Crear una serie a partir de una lista
datos = [10, 20, 30, 40, 50]
serie = pd.Series(datos)

print(serie)
```

In [11]:
import pandas as pd

# Crear una serie a partir de una lista
datos = [10, 20, 30, 40, 50]
serie = pd.Series(datos)

print(serie)

0    10
1    20
2    30
3    40
4    50
dtype: int64


## Accediendo a los elementos de una serie usando el índice
```python
import pandas as pd
# Crear una serie con índices personalizados
datos = [10, 20, 30, 40, 50]
serie = pd.Series(datos)
print(serie[0])  # Accede al elemento con índice 0
print(serie[1])  # Accede al elemento con índice 1
print(serie[2])  # Accede al elemento con índice 2
print(serie[3])  # Accede al elemento con índice 3
print(serie[4])  # Accede al elemento con índice 4
```

In [12]:
import pandas as pd
# Crear una serie con índices personalizados
datos = [10, 20, 30, 40, 50]
serie = pd.Series(datos)
print(serie[0])  # Accede al elemento con índice 0
print(serie[1])  # Accede al elemento con índice 1
print(serie[2])  # Accede al elemento con índice 2
print(serie[3])  # Accede al elemento con índice 3
print(serie[4])  # Accede al elemento con índice 4

10
20
30
40
50


## Podemos cambiar los índices de una serie pandas agregando el parámetro `index` al momento de crear la serie

```python
import pandas as pd   
# Crear una serie con índices personalizados
datos = ["Pera", "Manzana", "Naranja", "Plátano", "Uva"]
serie1 = pd.Series(datos, index=['a', 'b', 'c', 'd', 'e']) 
print(serie)
```

In [13]:
import pandas as pd   
# Crear una serie con índices personalizados
datos = ["Pera", "Manzana", "Naranja", "Plátano", "Uva"]
serie1 = pd.Series(datos, index=['a', 'b', 'c', 'd', 'e']) 
print(serie)

0    10
1    20
2    30
3    40
4    50
dtype: int64


## Acceder a los elementos de una serie usando sus ínidces personalizados
```python
print(serie1['a'])  # Accede al elemento con índice 'a'
print(serie1['b'])  # Accede al elemento con índice 'b'
print(serie1['c'])  # Accede al elemento con índice 'c'
print(serie1['d'])  # Accede al elemento con índice 'd'
print(serie1['e'])  # Accede al elemento con índice 'e'
```

In [14]:
print(serie1['a'])  # Accede al elemento con índice 'a'
print(serie1['b'])  # Accede al elemento con índice 'b'
print(serie1['c'])  # Accede al elemento con índice 'c'
print(serie1['d'])  # Accede al elemento con índice 'd'
print(serie1['e'])  # Accede al elemento con índice 'e'

Pera
Manzana
Naranja
Plátano
Uva


## Ordenando una serie pandas
```python
import pandas as pd
# Crear una serie con datos desordenados          
datos = [50, 20, 10, 40, 30]
serie = pd.Series(datos)
# Ordenar la serie
serie_ordenada = serie.sort_values()
print(serie_ordenada)
```

In [15]:
import pandas as pd
# Crear una serie con datos desordenados          
datos = [50, 20, 10, 40, 30]
serie = pd.Series(datos)
# Ordenar la serie
serie_ordenada = serie.sort_values()
print(serie_ordenada)

2    10
1    20
4    30
3    40
0    50
dtype: int64


## Ordenando de manera descendente
```   python
serie_ordenada_desc = serie.sort_values(ascending=False)
print(serie_ordenada_desc)
```

In [16]:
serie_ordenada_desc = serie.sort_values(ascending=False)
print(serie_ordenada_desc)

0    50
3    40
4    30
1    20
2    10
dtype: int64


## Cálculo de máximos, mínimos, suma, desviación estandard.        

```python
print(serie.max())  # Máximo valor de la serie
print(serie.min())  # Mínimo valor de la serie
print(serie.sum())  # Suma de la serie
print(serie.max())  # Máximo de la serie
print(serie.std())  # desviación estándar     
```



In [17]:
print(serie.max())  # Máximo valor de la serie
print(serie.min())  # Mínimo valor de la serie
print(serie.sum())  # Suma de la serie
print(serie.max())  # Máximo de la serie
print(serie.std())  # desviación estándar

50
10
150
50
15.811388300841896


## Selección de elementos en un dataframe

## Selección de filas con el método `.loc`


In [18]:
import pandas as pd

data = {
  "time":[120, 60, 45],
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}
df = pd.DataFrame(data)

print(df) 


   time  calories  duration
0   120       420        50
1    60       380        40
2    45       390        45


In [19]:
# seleccionando la primera fila 
# Si se usa un solo corchete retorna una serie 
print(df.iloc[0]) 
print(type(df.iloc[0]))

time        120
calories    420
duration     50
Name: 0, dtype: int64
<class 'pandas.core.series.Series'>


In [20]:
# Seleccionando la primera fila (otra forma)
# si se usan dos corchetes retorna un dataframe
print(df.iloc[[0]])
print(type(df.iloc[[0]]))

   time  calories  duration
0   120       420        50
<class 'pandas.core.frame.DataFrame'>


In [21]:
# Seleccionando varias filas
print(df.iloc[[1, 2]])

   time  calories  duration
1    60       380        40
2    45       390        45


## Nombrando los índices de un dataframe



In [22]:
data = {
  "time":[120, 60, 45, 60],
  "calories": [420, 380, 390, 400],
  "duration": [50, 40, 45, 39]
}
df = pd.DataFrame(data, index=["day1", "day2", "day3", "day4"])

print(df) 

      time  calories  duration
day1   120       420        50
day2    60       380        40
day3    45       390        45
day4    60       400        39


## Seleccionando filas con nombres con .loc

In [23]:
df.loc["day1"] # Esto es una serie

time        120
calories    420
duration     50
Name: day1, dtype: int64

In [24]:
df.loc[["day1"]] # Esto es un dataframe

time  calories  duration
day1   120       420        50

In [25]:
df.loc[["day2", "day4", "day3"]]

time  calories  duration
day2    60       380        40
day4    60       400        39
day3    45       390        45

## Recuerda usa `.iloc` para índices numéricos y `.loc` con índices con nombres.

## Se pueden usar rangos para seleccionar filas

In [26]:
df.loc["day1":"day3"] # selecciona desde el primer índice hasta el último, ambos extremos incluidos.

time  calories  duration
day1   120       420        50
day2    60       380        40
day3    45       390        45

## Clase 3

## Selección de columnas en un dataframe


In [27]:
import pandas as pd


df = pd.DataFrame({
    'Nombre': ['Ana', 'Luis', 'Pedro', 'Marta', 'Josefa', 'Matías', 'José', 'Felipe', 'Raúl', 'Claudia'],
    'Apellido': ['Cid', 'Rodriguez', 'Poblete', 'Valenzuela', 'Rodriguez', 'Valencia', 'Gómez', 'Soto', 'Muñoz', 'Herrera'],
    'Edad': [20, 25, 30, 23, 16, 28, 35, 22, 27, 31], 
    'Ciudad': ['Santiago', 'Valparaíso', 'Concepción', 'Rancagua', 'Rancagua', 'Temuco', 'Antofagasta', 'La Serena', 'Iquique', 'Puerto Montt'],
    'Soltero': [True, False, True, True, False, True, False, True, False, True],
    'Carrera': ['Medicina','Ingeniería','Arquitectura','Pedagogía',
    'Ingeniería','Medicina','Pedagogía','Arquitectura','Ingeniería','Medicina']

})


df

Nombre    Apellido  Edad        Ciudad  Soltero       Carrera
0      Ana         Cid    20      Santiago     True      Medicina
1     Luis   Rodriguez    25    Valparaíso    False    Ingeniería
2    Pedro     Poblete    30    Concepción     True  Arquitectura
3    Marta  Valenzuela    23      Rancagua     True     Pedagogía
4   Josefa   Rodriguez    16      Rancagua    False    Ingeniería
5   Matías    Valencia    28        Temuco     True      Medicina
6     José       Gómez    35   Antofagasta    False     Pedagogía
7   Felipe        Soto    22     La Serena     True  Arquitectura
8     Raúl       Muñoz    27       Iquique    False    Ingeniería
9  Claudia     Herrera    31  Puerto Montt     True      Medicina

In [28]:
print(df['Nombre'])  # Devuelve una Serie con los nombres

0        Ana
1       Luis
2      Pedro
3      Marta
4     Josefa
5     Matías
6       José
7     Felipe
8       Raúl
9    Claudia
Name: Nombre, dtype: object


In [29]:
print(df[['Nombre']])  # Devuelve un dataframe con los nombres

    Nombre
0      Ana
1     Luis
2    Pedro
3    Marta
4   Josefa
5   Matías
6     José
7   Felipe
8     Raúl
9  Claudia


In [30]:
print(df[["Nombre", "Soltero"]]) # Seleccionando varias columnas

    Nombre  Soltero
0      Ana     True
1     Luis    False
2    Pedro     True
3    Marta     True
4   Josefa    False
5   Matías     True
6     José    False
7   Felipe     True
8     Raúl    False
9  Claudia     True


## Seleccionado con condiciones

In [31]:
# Personas mayortes de 25 años
print(df[df["Edad"]>25])

    Nombre  Apellido  Edad        Ciudad  Soltero       Carrera
2    Pedro   Poblete    30    Concepción     True  Arquitectura
5   Matías  Valencia    28        Temuco     True      Medicina
6     José     Gómez    35   Antofagasta    False     Pedagogía
8     Raúl     Muñoz    27       Iquique    False    Ingeniería
9  Claudia   Herrera    31  Puerto Montt     True      Medicina


In [32]:
# Seleccionando a la persona de nombre "Ana"
print(df[df["Nombre"]=="Ana"]) 

  Nombre Apellido  Edad    Ciudad  Soltero   Carrera
0    Ana      Cid    20  Santiago     True  Medicina


In [33]:
# Seleccionado a las personas de apellido 'Rodriguez'ArithmeticError
df[df['Apellido'] == 'Rodriguez']

Nombre   Apellido  Edad      Ciudad  Soltero     Carrera
1    Luis  Rodriguez    25  Valparaíso    False  Ingeniería
4  Josefa  Rodriguez    16    Rancagua    False  Ingeniería

In [34]:
# Seleccionando personas de apellido 'Rodriguez' que viven en 'Valparaiso'
df[(df['Apellido'] =='Rodriguez') & (df['Ciudad'] =='Valparaíso')]

Nombre   Apellido  Edad      Ciudad  Soltero     Carrera
1   Luis  Rodriguez    25  Valparaíso    False  Ingeniería

In [35]:
# Calculando el promedio de las edades

df['Edad'].mean()

25.7

In [36]:
# ¿Cuántas personas están solteras?
df["Soltero"].sum()

6

## Agregando columnas a un dataframe

In [37]:
# Agregar la columna 'Estatus' según la la lista ['A', 'S', 'S', 'A','A', 'A', 'S', 'S', 'A','A']
# A= Activo, S = Suspendido
df['Estatus'] = ['A', 'S', 'S', 'A','A', 'A', 'S', 'S', 'A','A']
df

Nombre    Apellido  Edad        Ciudad  Soltero       Carrera Estatus
0      Ana         Cid    20      Santiago     True      Medicina       A
1     Luis   Rodriguez    25    Valparaíso    False    Ingeniería       S
2    Pedro     Poblete    30    Concepción     True  Arquitectura       S
3    Marta  Valenzuela    23      Rancagua     True     Pedagogía       A
4   Josefa   Rodriguez    16      Rancagua    False    Ingeniería       A
5   Matías    Valencia    28        Temuco     True      Medicina       A
6     José       Gómez    35   Antofagasta    False     Pedagogía       S
7   Felipe        Soto    22     La Serena     True  Arquitectura       S
8     Raúl       Muñoz    27       Iquique    False    Ingeniería       A
9  Claudia     Herrera    31  Puerto Montt     True      Medicina       A

In [38]:
## Agregando Notas "P1", "P2", "P3"
df["P1"]=[4, 5, 6, 5, 4, 3, 2, 3, 2, 4]
df["P2"]=[3, 4, 5, 4, 6, 7, 3, 3, 5, 4]
df["P3"]=[2, 4, 6, 3, 4, 3, 2, 3, 2, 4]
df


Nombre    Apellido  Edad        Ciudad  Soltero       Carrera Estatus  P1  \
0      Ana         Cid    20      Santiago     True      Medicina       A   4   
1     Luis   Rodriguez    25    Valparaíso    False    Ingeniería       S   5   
2    Pedro     Poblete    30    Concepción     True  Arquitectura       S   6   
3    Marta  Valenzuela    23      Rancagua     True     Pedagogía       A   5   
4   Josefa   Rodriguez    16      Rancagua    False    Ingeniería       A   4   
5   Matías    Valencia    28        Temuco     True      Medicina       A   3   
6     José       Gómez    35   Antofagasta    False     Pedagogía       S   2   
7   Felipe        Soto    22     La Serena     True  Arquitectura       S   3   
8     Raúl       Muñoz    27       Iquique    False    Ingeniería       A   2   
9  Claudia     Herrera    31  Puerto Montt     True      Medicina       A   4   

   P2  P3  
0   3   2  
1   4   4  
2   5   6  
3   4   3  
4   6   4  
5   7   3  
6   3   2  
7   3   3  
8   5   2  
9   4   4

In [39]:
## Agregando promedio de los estudiantes
#df["Promedio"] = (df["P1"]+df["P2"]+df["P3"])/3
# Otra forma más automática es 
df["Promedio"] = df[['P1', 'P2', 'P3']].mean(axis=1)

df

Nombre    Apellido  Edad        Ciudad  Soltero       Carrera Estatus  P1  \
0      Ana         Cid    20      Santiago     True      Medicina       A   4   
1     Luis   Rodriguez    25    Valparaíso    False    Ingeniería       S   5   
2    Pedro     Poblete    30    Concepción     True  Arquitectura       S   6   
3    Marta  Valenzuela    23      Rancagua     True     Pedagogía       A   5   
4   Josefa   Rodriguez    16      Rancagua    False    Ingeniería       A   4   
5   Matías    Valencia    28        Temuco     True      Medicina       A   3   
6     José       Gómez    35   Antofagasta    False     Pedagogía       S   2   
7   Felipe        Soto    22     La Serena     True  Arquitectura       S   3   
8     Raúl       Muñoz    27       Iquique    False    Ingeniería       A   2   
9  Claudia     Herrera    31  Puerto Montt     True      Medicina       A   4   

   P2  P3  Promedio  
0   3   2  3.000000  
1   4   4  4.333333  
2   5   6  5.666667  
3   4   3  4.000000  
4   6   4  4.666667  
5   7   3  4.333333  
6   3   2  2.333333  
7   3   3  3.000000  
8   5   2  3.000000  
9   4   4  4.000000

In [40]:
# Mostrando solo 2 decimales en promedio
pd.options.display.float_format = '{:.2f}'.format  # Muestra con 2 decimales
df

Nombre    Apellido  Edad        Ciudad  Soltero       Carrera Estatus  P1  \
0      Ana         Cid    20      Santiago     True      Medicina       A   4   
1     Luis   Rodriguez    25    Valparaíso    False    Ingeniería       S   5   
2    Pedro     Poblete    30    Concepción     True  Arquitectura       S   6   
3    Marta  Valenzuela    23      Rancagua     True     Pedagogía       A   5   
4   Josefa   Rodriguez    16      Rancagua    False    Ingeniería       A   4   
5   Matías    Valencia    28        Temuco     True      Medicina       A   3   
6     José       Gómez    35   Antofagasta    False     Pedagogía       S   2   
7   Felipe        Soto    22     La Serena     True  Arquitectura       S   3   
8     Raúl       Muñoz    27       Iquique    False    Ingeniería       A   2   
9  Claudia     Herrera    31  Puerto Montt     True      Medicina       A   4   

   P2  P3  Promedio  
0   3   2      3.00  
1   4   4      4.33  
2   5   6      5.67  
3   4   3      4.00  
4   6   4      4.67  
5   7   3      4.33  
6   3   2      2.33  
7   3   3      3.00  
8   5   2      3.00  
9   4   4      4.00

## 

In [41]:
#  Clasificar estudiantes, agregando la columnna clasificación mostrando "A" si el estudiante aprobó y "R" si el estudiante no aprobó

def clasificacion(nota):
    if nota >= 4:
        return "A"
    else:
        return "B"
df["Clasificacion"]= df["Promedio"].apply(clasificacion)
df


Nombre    Apellido  Edad        Ciudad  Soltero       Carrera Estatus  P1  \
0      Ana         Cid    20      Santiago     True      Medicina       A   4   
1     Luis   Rodriguez    25    Valparaíso    False    Ingeniería       S   5   
2    Pedro     Poblete    30    Concepción     True  Arquitectura       S   6   
3    Marta  Valenzuela    23      Rancagua     True     Pedagogía       A   5   
4   Josefa   Rodriguez    16      Rancagua    False    Ingeniería       A   4   
5   Matías    Valencia    28        Temuco     True      Medicina       A   3   
6     José       Gómez    35   Antofagasta    False     Pedagogía       S   2   
7   Felipe        Soto    22     La Serena     True  Arquitectura       S   3   
8     Raúl       Muñoz    27       Iquique    False    Ingeniería       A   2   
9  Claudia     Herrera    31  Puerto Montt     True      Medicina       A   4   

   P2  P3  Promedio Clasificacion  
0   3   2      3.00             B  
1   4   4      4.33             A  
2   5   6      5.67             A  
3   4   3      4.00             A  
4   6   4      4.67             A  
5   7   3      4.33             A  
6   3   2      2.33             B  
7   3   3      3.00             B  
8   5   2      3.00             B  
9   4   4      4.00             A

## Ejercicios
- Muestra sólo los aprobados en pantalla
- Muestra sólo los reprobados
- Elimina las columnas Ciudad, Estatus, Soltero del dataframe original
- Muestra en pantalla los estudiantes aprobados mayores de 30 años
- Muestra en pantalla los estudiantes menore de 30 reprobados y calcula su promedio
- Calcula el promedio de los estudiantes de medicina
- Calcula cuántos estudiantes aprobaron
- ¿Qué podemos decir de los estudiantes de ingeniería?

## Clase 4

## Eliminar columnas de un dataframe
Para eliminar columnnas se usa el método `.drop()` con la siguiente sintaxis

1. ```df.drop(columns=['column_name1', 'column_name2', ....])```: Este método retorna un dataframe, pero no elimina las columnas del dataframe original.
2. Para eliminar efectivamente las columnnas del dataframe original se debe agregar el argumento ```inplace=True```.


In [42]:
# Mostremos las columnas del dataframe
print(df.columns)

Index(['Nombre', 'Apellido', 'Edad', 'Ciudad', 'Soltero', 'Carrera', 'Estatus',
       'P1', 'P2', 'P3', 'Promedio', 'Clasificacion'],
      dtype='object')


In [43]:
df.drop(columns=["Estatus"])

Nombre    Apellido  Edad        Ciudad  Soltero       Carrera  P1  P2  P3  \
0      Ana         Cid    20      Santiago     True      Medicina   4   3   2   
1     Luis   Rodriguez    25    Valparaíso    False    Ingeniería   5   4   4   
2    Pedro     Poblete    30    Concepción     True  Arquitectura   6   5   6   
3    Marta  Valenzuela    23      Rancagua     True     Pedagogía   5   4   3   
4   Josefa   Rodriguez    16      Rancagua    False    Ingeniería   4   6   4   
5   Matías    Valencia    28        Temuco     True      Medicina   3   7   3   
6     José       Gómez    35   Antofagasta    False     Pedagogía   2   3   2   
7   Felipe        Soto    22     La Serena     True  Arquitectura   3   3   3   
8     Raúl       Muñoz    27       Iquique    False    Ingeniería   2   5   2   
9  Claudia     Herrera    31  Puerto Montt     True      Medicina   4   4   4   

   Promedio Clasificacion  
0      3.00             B  
1      4.33             A  
2      5.67             A  
3      4.00             A  
4      4.67             A  
5      4.33             A  
6      2.33             B  
7      3.00             B  
8      3.00             B  
9      4.00             A

In [44]:
# Dataframe original (no alterado)
df

Nombre    Apellido  Edad        Ciudad  Soltero       Carrera Estatus  P1  \
0      Ana         Cid    20      Santiago     True      Medicina       A   4   
1     Luis   Rodriguez    25    Valparaíso    False    Ingeniería       S   5   
2    Pedro     Poblete    30    Concepción     True  Arquitectura       S   6   
3    Marta  Valenzuela    23      Rancagua     True     Pedagogía       A   5   
4   Josefa   Rodriguez    16      Rancagua    False    Ingeniería       A   4   
5   Matías    Valencia    28        Temuco     True      Medicina       A   3   
6     José       Gómez    35   Antofagasta    False     Pedagogía       S   2   
7   Felipe        Soto    22     La Serena     True  Arquitectura       S   3   
8     Raúl       Muñoz    27       Iquique    False    Ingeniería       A   2   
9  Claudia     Herrera    31  Puerto Montt     True      Medicina       A   4   

   P2  P3  Promedio Clasificacion  
0   3   2      3.00             B  
1   4   4      4.33             A  
2   5   6      5.67             A  
3   4   3      4.00             A  
4   6   4      4.67             A  
5   7   3      4.33             A  
6   3   2      2.33             B  
7   3   3      3.00             B  
8   5   2      3.00             B  
9   4   4      4.00             A

In [45]:
 # Usando el argumento inplace
df.drop(columns=["Estatus", "Soltero"], inplace=True)

In [46]:
# Dataframe original (alterado)
df

Nombre    Apellido  Edad        Ciudad       Carrera  P1  P2  P3  \
0      Ana         Cid    20      Santiago      Medicina   4   3   2   
1     Luis   Rodriguez    25    Valparaíso    Ingeniería   5   4   4   
2    Pedro     Poblete    30    Concepción  Arquitectura   6   5   6   
3    Marta  Valenzuela    23      Rancagua     Pedagogía   5   4   3   
4   Josefa   Rodriguez    16      Rancagua    Ingeniería   4   6   4   
5   Matías    Valencia    28        Temuco      Medicina   3   7   3   
6     José       Gómez    35   Antofagasta     Pedagogía   2   3   2   
7   Felipe        Soto    22     La Serena  Arquitectura   3   3   3   
8     Raúl       Muñoz    27       Iquique    Ingeniería   2   5   2   
9  Claudia     Herrera    31  Puerto Montt      Medicina   4   4   4   

   Promedio Clasificacion  
0      3.00             B  
1      4.33             A  
2      5.67             A  
3      4.00             A  
4      4.67             A  
5      4.33             A  
6      2.33             B  
7      3.00             B  
8      3.00             B  
9      4.00             A

## Eliminando filas de dataframe
1. Eliminado por índice


In [47]:
import pandas as pd

# Crear un DataFrame de ejemplo
df = pd.DataFrame({
    "Nombre": ["Ana", "Luis", "Pedro"],
    "Edad": [23, 35, 40]
}, index=["a", "b", "c"])

# Mostrar df

print(df)

  Nombre  Edad
a    Ana    23
b   Luis    35
c  Pedro    40


In [48]:
# Eliminar la fila con índice "b"
df_sin_b = df.drop("b")
print(df_sin_b)

  Nombre  Edad
a    Ana    23
c  Pedro    40


2. Eliminado de varias filas por índice 

In [49]:

df_sin_varias = df.drop(["a", "c"])


3. Eliminar filas de dataframe original

In [50]:
df_sin_varias = df.drop(["a", "c"], inplace=True)
print(df)

  Nombre  Edad
b   Luis    35


4. Eliminado de filas con índices numéricos por default

In [51]:

import pandas as pd

# Crear un DataFrame de ejemplo
df = pd.DataFrame({
    "Nombre": ["Ana", "Luis", "Pedro", "marcos", "carlos"],
    "Edad": [23, 35, 40, 30, 20]
})

# Mostrar df

print(df)
df.drop([1,2,3]) # Elimina filas 1, 2, 3


   Nombre  Edad
0     Ana    23
1    Luis    35
2   Pedro    40
3  marcos    30
4  carlos    20


Nombre  Edad
0     Ana    23
4  carlos    20

5. Eliminado por condición (filtrado)
```python
# Eliminar filas donde Edad es mayor de 30
df_filtrado = df[df["Edad"] <= 30]

```

## Leer archivo con formato JSON en Pandas

In [52]:
# Leyendo con pd.read_json()
import pandas as pd

df = pd.read_json('data.json')

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130    409.10
1          60    117       145    479.00
2          60    103       135    340.00
3          45    109       175    282.40
4          45    117       148    406.00
5          60    102       127    300.50
6          60    110       136    374.00
7          45    104       134    253.30
8          30    109       133    195.10
9          60     98       124    269.00
10         60    103       147    329.30
11         60    100       120    250.70
12         60    106       128    345.30
13         60    104       132    379.30
14         60     98       123    275.00
15         60     98       120    215.20
16         60    100       120    300.00
17         45     90       112       NaN
18         60    103       123    323.00
19         45     97       125    243.00
20         60    108       131    364.20
21         45    100       119    282.00
22         60    130       101    300.00
23         45   

## Creando archivo json en python 
JSON = Dicionario en python

In [53]:
# Creando archivo json en python 
import pandas as pd

data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "1":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409.1,
    "1":479.0,
    "2":340.0,
    "3":282.4,
    "4":406.0,
    "5":300.5
  }
}

df = pd.DataFrame(data)

print(df) 


   Duration  Pulse  Maxpulse  Calories
0        60    110       130    409.10
1        60    117       145    479.00
2        60    103       135    340.00
3        45    109       175    282.40
4        45    117       148    406.00
5        60    102       127    300.50


# Clase 5
## Limpieza de datos
La limpieza de datos significa corregir los datos incorrectos en tu conjunto de datos.

Los datos incorrectos pueden ser:

- Celdas vacías
- Datos en formato incorrecto
- Datos erróneos
- Duplicados

Vamos a trabajar con el conjuntos de datos llamado `datos_crudos.csv`, los cuales se muestran a continuación:


In [54]:
import pandas as pd

# Importando archivo
df = pd.read_csv('datos_crudos.csv')
# Mostrando archivo completo en pantalla
print(df.to_string)

<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120  

### Podemos observar:
- El conjunto de datos contiene algunas celdas vacías ("Fecha" en la fila 22, y "Calorías" en las filas 18 y 28).
- El conjunto de datos contiene un formato incorrecto ("Fecha" en la fila 26).
- El conjunto de datos contiene datos erróneos ("Duración" en la fila 7).
- El conjunto de datos contiene duplicados (filas 11 y 12).

### 🟨 Celdas Vacías
Las celdas vacías pueden ocasionar resultados incorrectos al analizar los datos.
### ✂️ Eliminar Filas
Una forma de abordar las celdas vacías es eliminar las filas que las contienen.
Esto suele estar bien, ya que los conjuntos de datos pueden ser muy grandes, y eliminar unas pocas filas no tendrá un gran impacto en el resultado final.


In [55]:
import pandas as pd

df = pd.read_csv('datos_crudos.csv')

new_df = df.dropna()

print(new_df.to_string())

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'  

### 📌 Nota: Por defecto, el método `dropna()` en pandas devuelve un nuevo DataFrame, sin modificar el original.
### 🔧 Si deseas que los cambios se apliquen directamente al DataFrame original, puedes usar el parámetro `inplace=True`


In [56]:
import pandas as pd

df = pd.read_csv('datos_crudos.csv')

df.dropna(inplace = True)

print(df.to_string())

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'  

### 📌 Nota: Ahora, el método `dropna(inplace=True)` no devuelve un nuevo DataFrame, sino que elimina directamente todas las filas que contienen valores nulos (NULL) del DataFrame original.
### 📉 Es una forma útil de limpiar datos cuando ya no necesitas conservar la versión original. Ten en cuenta que este cambio es permanente en la variable utilizada, a menos que se haya hecho una copia previamente.


## 🔄 Reemplazar Valores Vacíos
- Otra forma de manejar las celdas vacías es insertar un nuevo valor en lugar de eliminarlas.
- De este modo, no es necesario borrar filas completas solo por algunas celdas vacías.
- 📚 El método fillna() en pandas permite reemplazar las celdas vacías (NaN) con un valor que tú elijas:


In [57]:
# Reemplazando los valores nulos con el valor 130
import pandas as pd

df = pd.read_csv('datos_crudos.csv')

df.fillna(130, inplace = True)
print(df.to_string)

<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120  

## 🔹 Reemplazar solo en columnas específicas
- El ejemplo anterior reemplaza todas las celdas vacías en todo el DataFrame.
- Para reemplazar los valores vacíos únicamente en una columna, especifica el nombre de la columna dentro del DataFrame.


In [58]:
#  Reemplazar valores NULL en la columna "Calorías" con el número 130:

import pandas as pd

df = pd.read_csv('datos_crudos.csv')

df.fillna({"Calories": 130}, inplace=True)

print(df.to_string)


<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120  

## 🔹 Reemplazar usando la media, la mediana o la moda
- Una forma común de reemplazar celdas vacías es calcular el valor de media, mediana o moda de la columna.
- 📊 Pandas utiliza los métodos `mean()`, `median()` y `mode()` para calcular estos valores en la columna especificada.



In [59]:
# Calcular la MEDIA y reemplazar cualquier valor vacío con ella:

import pandas as pd

df = pd.read_csv('datos_crudos.csv')

x = df["Calories"].mean()

df.fillna({"Calories": x}, inplace=True)

print(df.to_string)

<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120  

In [60]:
# Calcular la MEDIANA y reemplazar cualquier valor vacío con ella:

import pandas as pd

df = pd.read_csv('datos_crudos.csv')

x = df["Calories"].median()

df.fillna({"Calories": x}, inplace=True)

print(df.to_string)

<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120  

In [61]:
# Calcular la MODA y reemplazar cualquier valor vacío con ella:
import pandas as pd

df = pd.read_csv('datOS_crudos.csv')

x = df["Calories"].mode()[0]

df.fillna({"Calories": x}, inplace=True)
print(df.to_string)

<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120  

## 🔹 Datos con formato incorrecto
Las celdas con datos en un formato erróneo pueden dificultar, o incluso imposibilitar, el análisis de los datos.
### 📌 Para solucionarlo, tienes dos opciones:
- Eliminar las filas que contienen dichos errores.
- Convertir todas las celdas de la columna al mismo formato.
### Convertir a un formato correcto
- En nuestro DataFrame, hay dos celdas con formato incorrecto. 
- Revisa las filas 22 y 26: la columna 'Fecha' debería contener una cadena que represente una fecha.




In [62]:
import pandas as pd
df = pd.read_csv('datoS_crudos.csv')
print(df.to_string)

<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120  

### Intentemos convertir todas las celdas de la columna 'Fecha' en fechas. 
Pandas tiene un método llamado `to_datetime()` para hacerlo:


In [63]:
import pandas as pd

df = pd.read_csv('datos_crudos.csv')

df['Date'] = pd.to_datetime(df['Date'], format='mixed')

print(df.to_string())


    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130    409.10
1         60 2020-12-02    117       145    479.00
2         60 2020-12-03    103       135    340.00
3         45 2020-12-04    109       175    282.40
4         45 2020-12-05    117       148    406.00
5         60 2020-12-06    102       127    300.00
6         60 2020-12-07    110       136    374.00
7        450 2020-12-08    104       134    253.30
8         30 2020-12-09    109       133    195.10
9         60 2020-12-10     98       124    269.00
10        60 2020-12-11    103       147    329.30
11        60 2020-12-12    100       120    250.70
12        60 2020-12-12    100       120    250.70
13        60 2020-12-13    106       128    345.30
14        60 2020-12-14    104       132    379.30
15        60 2020-12-15     98       123    275.00
16        60 2020-12-16     98       120    215.20
17        60 2020-12-17    100       120    300.00
18        45 2020-12-18     90 

🔎 Como puedes ver en el resultado, la fecha en la fila 26 fue corregida, pero la fecha vacía en la fila 22 recibió un valor NaT (Not a Time), es decir, un valor vacío. Una forma de manejar estos valores vacíos es simplemente eliminar toda la fila.




### Eliminar filas
El resultado de la conversión en el ejemplo anterior nos dio un valor NaT, que puede ser tratado como un valor NULL, y podemos eliminar la fila utilizando el método dropna().



In [64]:
# Eliminar filas con un valor NULL en la columna "Fecha":
import pandas as pd

df = pd.read_csv('datos_crudos.csv')
df.dropna(subset=['Date'], inplace = True)
print(df.to_string)



<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120  

## Datos erróneos
- Los "datos incorrectos" no tienen por qué ser "celdas vacías" o "formato incorrecto", pueden simplemente se equivoca, como si alguien registrara "199" en lugar de "1.99".

- A veces se pueden detectar datos erróneos mirando el conjunto de datos, porque se tiene una expectativa de lo que Debería serlo.

- Si echas un vistazo a nuestro conjunto de datos, puedes ver que en la fila 7, la duración es de 450, pero para todas las demás filas la duración está entre 30 y 60.

- No tiene por qué estar mal, pero considerando que este es el conjunto de datos de las sesiones de entrenamiento de alguien, concluimos que esta persona no entrenó durante 450 minutos.
- ¿Cómo pódemos correfir este proiblema?

In [65]:
import pandas as pd
df = pd.read_csv('datoS_crudos.csv')
print(df.to_string)

<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120  

### Sustitución de valores

In [66]:
# Establezca "Duración" = 45 en la fila 7:
df.loc[7, 'Duration'] = 45
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7         45  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'    100       120    300.00
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123    323.00
20        45  '2020/12/20'     97       125    243.00
21        60  '2020/12/21'    108       131    364.20
22        45           NaN    100       119    282.00
23        60  '2020/12/23'    130       101    300.00
24        45  '2020/12/24'    105       132    246.00
25        60  '2020/12/25'    102       126    334.50
26        60      20201226    100       120    250.00
27        60  '2020/12/27'     92       118    241.00
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132    280.00
30        60  '2020/12/30'    102       129    380.30
31        60  '2020/12/31'     92       115    243.00

### Observación
En el caso de conjuntos de datos pequeños, es posible que pueda reemplazar los datos incorrectos uno por uno, pero no para grandes conjuntos de datos.

Para reemplazar los datos incorrectos por conjuntos de datos más grandes, puede crear algunas reglas, p. ej. Establezca algunos límites para los valores legales y reemplace los valores que estén fuera de la propiedad Límites.

Ejemplo
- Recorra todos los valores de la columna "Duración".

- Si el valor es superior a 120, establézcalo en 120:

In [67]:
import pandas as pd
df = pd.read_csv('datoS_crudos.csv')
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.loc[x, "Duration"] = 120
  
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        120  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'    100       120    300.00
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123    323.00
20        45  '2020/12/20'     97       125    243.00
21        60  '2020/12/21'    108       131    364.20
22        45           NaN    100       119    282.00
23        60  '2020/12/23'    130       101    300.00
24        45  '2020/12/24'    105       132    246.00
25        60  '2020/12/25'    102       126    334.50
26        60      20201226    100       120    250.00
27        60  '2020/12/27'     92       118    241.00
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132    280.00
30        60  '2020/12/30'    102       129    380.30
31        60  '2020/12/31'     92       115    243.00

### Eliminación de filas con datos incorrectos
- Otra forma de controlar los datos incorrectos es eliminar las filas que contienen datos incorrectos.

- De esta manera, no tiene que averiguar con qué reemplazarlos, y es muy probable que no los necesites para hacer tus análisis.
- Ejemplo: Elimine las filas en las que "Duración" sea superior a 120:

In [76]:
import pandas as pd
df = pd.read_csv('datoS_crudos.csv')
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.drop(x, inplace = True)


df.isnull().sum()


Duration    0
Date        1
Pulse       0
Maxpulse    0
Calories    2
dtype: int64

## Eliminación de duplicados
- Las filas duplicadas son filas que se hayan registrado más de una vez.
- Al echar un vistazo a nuestro conjunto de datos de prueba, podemos suponer que las filas 11 y 12 están duplicadas.

- Para descubrir duplicados, podemos usar el método `.duplicated()`

- El método devuelve valores booleanos para cada fila.

In [69]:
import pandas as pd
df = pd.read_csv('datoS_crudos.csv')
print(df.to_string())

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
12        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'  

In [70]:
print(df.duplicated())
(df.duplicated()).sum()  # Cuenta el número de filas duplicadas

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


1

### Ejemplo de eliminación de duplicados
- Para eliminar duplicados, utilice el método `.drop_duplicates()`

In [71]:
# Eliminar todos los duplicados
import pandas as pd
df = pd.read_csv('datos_crudos.csv')
df.drop_duplicates(inplace=True)
df

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130    409.10
1         60  '2020/12/02'    117       145    479.00
2         60  '2020/12/03'    103       135    340.00
3         45  '2020/12/04'    109       175    282.40
4         45  '2020/12/05'    117       148    406.00
5         60  '2020/12/06'    102       127    300.00
6         60  '2020/12/07'    110       136    374.00
7        450  '2020/12/08'    104       134    253.30
8         30  '2020/12/09'    109       133    195.10
9         60  '2020/12/10'     98       124    269.00
10        60  '2020/12/11'    103       147    329.30
11        60  '2020/12/12'    100       120    250.70
13        60  '2020/12/13'    106       128    345.30
14        60  '2020/12/14'    104       132    379.30
15        60  '2020/12/15'     98       123    275.00
16        60  '2020/12/16'     98       120    215.20
17        60  '2020/12/17'    100       120    300.00
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123    323.00
20        45  '2020/12/20'     97       125    243.00
21        60  '2020/12/21'    108       131    364.20
22        45           NaN    100       119    282.00
23        60  '2020/12/23'    130       101    300.00
24        45  '2020/12/24'    105       132    246.00
25        60  '2020/12/25'    102       126    334.50
26        60      20201226    100       120    250.00
27        60  '2020/12/27'     92       118    241.00
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132    280.00
30        60  '2020/12/30'    102       129    380.30
31        60  '2020/12/31'     92       115    243.00

In [72]:
# Contando los duplicados
df.duplicated().sum()

0